lib/semigrp.gi

```diff
@@ -542,7 +542,25 @@ InstallMethod( AsSemigroup,
     "generic method for collections",
     [ IsCollection ],
     function ( D )
-    local   S,  L;
+    local L, S, i, j, k;
+
+    # Must check associativity
+    if ApplicableMethod(IsAssociative, [D]) <> fail then
```

Comment:
Just to be clear, the only "concern" I have about this PR is that I am wary of using code that relies on `ApplicableMethod` to make decisions.

However, I acknowledge this is a subjective feeling. So I don't want to block this PR over it; I just wanted to make sure we consider alternatives.

Reply:
I'm in no rush for this to be merged and I'll be happy to rework the change along the lines discussed above.

But this PR is currently a wrong-result bugfix, so if get close to releasing GAP 4.12.0 and I've still not made the further (non-bugfix) changes, then we should merge this as-is.